Profiles

tylerparsons opened the pull request:
# Overview

These changes encompass the entirety of support for the profiles framework as well as a suite of new features and workflow behavior governed by newly introduced profile attributes.

## Framework

The architecture and APIs for the profiles framework are documented in detail here: [datavis-cartodb/cartodb-docs/playbooks/profiles.md](https://bbgithub.dev.bloomberg.com/datavis-cartodb/cartodb-docs/blob/master/playbooks/profiles.md).  There are new ruby models, changes to the `User` model as well as a new `ProfileAttributes` BackBone model which serve as the primary interface for the profiles framework.

## New Profile Attributes Features

The new features for which the profiles framework is used include the following:

* Profile Enablement
  * Static enablement through `profiles_users` table
  * temporary or 'session' enablement via http headers
  * Rake Task for static enablement: `rake profiles:enable_user[username,profile_name]`
  * Rake Task for static disablement: `rake profiles:disable_user[username,profile_name]`
* Font
  * Updated Default Fixed Title Font
  * Updated Default Fixed Description Font
  * Support for a new set of allowed fonts as part of profile attributes
* Export Image
  * All styles transferred from fixed title/description to export image view
  * All styles including font and opacity transferred to actual exported image
* Locking Map
  * Lock Pan on a map
  * Lock Zoom on a map
* Default Overlays
  * Fixed title on by default flag
  * New fixed title default opacity
  * New fixed description default opacity
* Default basemaps
  * Optional whitelist of certain basemaps for which to enable user

Any feature behind a profile attribute will not change default behavior if the profile attribute is not set for the user.  Aside from some bug fixes, the workflow of user's without profile attributes will be unchanged.

tylerparsons committed:
Update dependencies for mac os x builds

tylerparsons committed:
Create framework for leveraging user profiles

Add support for storing 'profiles', their corresponding
'attrs', and for linking profiles to users through the
'profiles_users' table.

Profiles may also be linked to users temporarily through
the 'SessionProfileManager', which adds certain profiles
to a user scope session. All profiles to which a user is
linked are exposed through the 'ProfileAttributes' interface,
which merges the attributes of user and session profiles
into a single hash containing all profile attributes.

Profile attributes support use cases such as the configuring
of default overlays, which is exemplified in the 'Overlays'
class.

tylerparsons committed:
Bundle update

tylerparsons committed:
Expose lock pan and zoom as map properties

tylerparsons committed:
Backout profile attributes from visualizations view

tylerparsons committed:
Expose profile attributes through rest endpoint

tylerparsons committed:
Enable lock pan and zoom options by default

tylerparsons committed:
Preserve title overlay styles in export image helper

tylerparsons committed:
Reimplement session profiles using redis

* Move profiles implementation into user model
* Expose profile_attributes in static user data

tylerparsons committed:
Enable AVMM export option through profile attribute

tylerparsons committed:
Simplify read-only ProfileAttributes model

tylerparsons committed:
Update generated cartodb.js files

tylerparsons committed:
Preserve fixed title opacity and padding in export image view

tylerparsons committed:
Include profile_attributes.js as public maps dependency

tylerparsons committed:
Support news fonts in overlays

tylerparsons committed:
Add new fonts to copy task

tylerparsons committed:
Include old_common fonts in public map

tylerparsons committed:
Configure default fonts, deep copy font forms

tylerparsons committed:
Replace fonts, add configurable form width

tylerparsons committed:
Enforce default font in fixed title

tylerparsons committed:
Fix definition of certain font css classes

tylerparsons committed:
Support new default fixed title and desription font sizes

tylerparsons committed:
Use static ProfileAttributes load util

tylerparsons committed:
Re-enable AVMM export

tylerparsons committed:
Support transactional replacement of session profiles

tylerparsons committed:
Support profile enablements with http headers

tylerparsons committed:
Cleanup session profiles after deletion

tylerparsons committed:
Propagate additional styles to export image view

tylerparsons committed:
Support whitelisting basemaps with profile attributes

tylerparsons committed:
Add profiles enablement rake task

tylerparsons committed:
Fix overall opacity of export image overlays

tylerparsons committed:
Reset vendor files to blp_dev version

tylerparsons commented on app/controllers/api/json/profile_attributes_controller.rb:
```diff
@@ -0,0 +1,10 @@
+require_relative '../../../../lib/profile_attributes'
+
+class Api::Json::ProfileAttributesController < Api::ApplicationController
+  include CartoDB
```
Ensure that this request is authenticated i.e. `before_filter` contains `http_header_authentication`.

tylerparsons replied:
Confirmed - http_header_authentication is called by default in `Api::ApplicationController` unless a `skip_before_filter` is added in the sub controller. The request throws a 401 if the caller is not authenticated.

tylerparsons commented on lib/assets/javascripts/cartodb/table/table.js:
```diff
@@ -78,15 +78,37 @@ $(function() {
 
-      _(layers).each(function(catLayers, cat) {
-        _(catLayers).map(function(m) {
+      // Enforce set of whitelisted basemaps, if one is specified in profile attributes.
```
It might be better/needed to have this in the backend.  Will have to doublecheck if this is respected when creating a new map when the first basemap is not in the whitelist.

tylerparsons replied:
This whitelisting has been moved to the backend in subsequent commits.

tylerparsons committed:
Replicate profiles logic in carto/user

tylerparsons committed:
Fix and move basemap filtering to backend

Also impose on that user is enabled for saved
basemap on the front end.  This is necessary
because the layers representing basemaps are
constructed on the frontend, thus the correct
layer can only be generated on the frontend.

tylerparsons committed:
Update carto_assets dependencies

tylerparsons committed:
Freeze carto_assets dependencies with shrinkwrap

mbektash commented on Gemfile:
```diff
@@ -33,6 +33,7 @@ gem 'ejs',                     '~> 1.1.1'
 gem 'execjs',                  '~> 0.4' # Required by ejs
 
 gem 'net-ldap',                '0.11'
+gem 'net-telnet'
```
please check changes to this file and revert if not required

mbektash commented on Gemfile.lock:
```diff
@@ -433,4 +433,4 @@ DEPENDENCIES
 
 BUNDLED WITH
-   1.11.2
+   1.15.0
```
please check changes to this file and revert if not required

mbektash commented on app/views/admin/visualizations/show.html.erb:
```diff
@@ -12,7 +12,7 @@
     , config = <%= safe_js_object frontend_config %>
     , map_data = <%= safe_js_object @visualization.map.public_values.to_json %>
-    , basemaps = <%=raw Cartodb.config[:basemaps].present? ? Cartodb.config[:basemaps].to_json.html_safe: 'null' %>
+    , basemaps = <%=raw @visualization.user.basemaps.to_json.html_safe || 'null' %>
```
please test with public maps, can @visualization.user be ever nil?

tylerparsons replied:
Public maps uses a different template - will double check this for sample maps

tylerparsons replied:
This view and its `basemaps` are only used in the map editor view.  I verified that public maps and sample maps will still display the basemaps with which they are saved, irrespective of the basemaps that the viewer is enabled for.  This is the intended behavior.

When a user saves a copy of a sample map, however, their default basemap will be used, which is always one of the basemaps for which they are enabled.  Currently, basemaps are not copied when saving a copy of a sample map.  This logic remains the same.

mbektash commented on db/migrate/20170523172137_add_lock_flags_to_maps.rb:
```diff
@@ -0,0 +1,10 @@
+    end
+  end
+
+end
```
please check if downgrade is removing these columns

tylerparsons replied:
Downgrade works as expected.

mbektash commented on lib/assets/javascripts/cartodb/models/vis.js:
```diff
@@ -197,7 +197,7 @@ cdb.admin.Visualization = cdb.core.Model.extend({
     }, this);
 
-    this.map.set('id', this.get('map_id'));
+    this._fetchMap();
```
please check if this is still needed

tylerparsons replied:
This is needed to load the `lock_pan` and `lock_zoom` properties of the map.  Before this change, the map model was not actually loaded - following these id change triggers will actually reveal that fetch is never called.  This will ensure that these flags are respected when a map is shown.

mbektash commented on lib/assets/javascripts/cartodb/table/basemap/basemap_dropdown/basemap_dropdown.js:
```diff
@@ -395,6 +395,20 @@ cdb.admin.DropdownBasemap = cdb.ui.common.Dropdown.extend({
+    var enabledForSavedBaseLayer =
+      !!this.baseLayers.filter(bl => map.getBaseLayer().isEqual(bl)).length;
+
+    if (!enabledForSavedBaseLayer) {
```
It may be better to move this logic to mapview or other parent widget

tylerparsons committed:
Merge branch 'blp_dev' into blp_dev_profiles

mbektash commented:
Submodule carto_assets seems to be reverted to an old commit

mbektash commented on lib/profile_attributes.rb:
```diff
@@ -0,0 +1,26 @@
+  class ProfileAttributes
+
+    def self.load(user, session)
+      user_profiles = user.profiles
```
do you need to check for nil user?

tylerparsons replied:
These files should actually be removed - I used them previously for a slightly different accessor design for the profile attributes but this logic has now been migrated to the user model

mbektash commented on lib/tasks/profiles.rake:
```diff
@@ -0,0 +1,23 @@
+namespace :profiles do
+  namespace :enable do
```
why not just task with name enable_user?

tylerparsons replied:
The thinking is that we could have other methods of enablement e.g. org, multiple users, user query, etc. down the line if we need

tylerparsons replied:
As discussed with Mehmet, will make this change.

mbektash commented on python_requirements.txt:
```diff
@@ -3,6 +3,6 @@ argparse==1.2.1
 redis==2.4.9
 python-varnish==0.1.2
-gdal==1.10.0
+gdal==2.1.0
```
if not required please revert

tylerparsons replied:
Will double check and try to revert these and other version changes

tylerparsons committed:
Update submodule dependencies

tylerparsons committed:
Restructure and add disable task to profile rake tasks

tylerparsons committed:
Fix profiles disable task typo

tylerparsons committed:
Prevent unnecessary round trip in profiles user task

tylerparsons committed:
Revert unnecessary python dependency changes

tylerparsons committed:
Revert unneeded Gemfile changes

tylerparsons committed:
Remove deprecated ProfileAttributes utility

tylerparsons committed:
Refactor default base layer selection

tylerparsons committed:
Bump profile migration version; fix cat fk down

tylerparsons committed:
Fix profiles_user migration name

tylerparsons committed:
Add profiles_users primary key

tylerparsons commented:
I've addressed all the comments and tested the changes.  If all looks good to @mbektas and the headless tests are successful, I think this PR is ready to merge.

tylerparsons commented on lib/carto/http_header_authentication.rb:
```diff
@@ -88,5 +92,30 @@ def header_value(headers)
       puts "user-auto-creation : Trying to extract value from headers for #{header}, value is #{headers[header]}"
       !header.nil? && !header.empty? ? headers[header] : nil
     end
+
```
@vishalya please review these changes to the authentication system when you get a chance

vishalya reviewed (commented):
Looks good.

mbektash commented on lib/assets/javascripts/cartodb/models/vis.js:
```diff
@@ -197,7 +197,7 @@ cdb.admin.Visualization = cdb.core.Model.extend({
     }, this);
 
-    this.map.set('id', this.get('map_id'));
+    this._fetchMap();
```
this._fetchMap(); is causing http://host/user/username/api/v1/maps to be called which returns 404. happening during New Map creation

tylerparsons replied:
Looking into this now.

tylerparsons replied:
Fixed with newest commit.  Tested for "New Map", "Open in Maps" with one or more datasets and sample maps "Save A Copy"

tylerparsons commented:
An asset version bump will also be required with these changes. The current blp_prod version is 3.23.72

tylerparsons committed:
Prevent map fetch when map_id unset

tylerparsons committed:
Increment asset version

tylerparsons commented:
Confirmed with product that we do want to keep lock pan and zoom turned on for everyone.

tylerparsons commented:
Closing. Will open a new pull request for blp_dev.

tylerparsons commented:
Approved by Mehmet and Vishal.  Will create a new PR for blp_dev.
